### ISSUE TITLE
NVCaffe 0.16 cudnn convolution selection and RAM consumption

### ISSUE THREAD

**Issue by RSly** [NVIDIA/caffe#341]
Hi @drnikolaev, I am opening this issue regarding the cudnn fastest algorithm selection and ram usage

Currently: 

> CuDNN  v6 comes packed with good convolution algorithms, and NVCaffe tries to choose the fastest one. Before trying, it checks how much memory would every candidate need to do the job and allocates one workspace for them all (in your case it's 7GB). Then it starts short races and choose the winner. If winner needs less than 7GB NVCaffe shrinks the workspace to the size the winner wants.

The problem I see is that in many cases we will end up either under-using the available gpu memory, or even worst end up with an out-of-memory crash. 
I explain the two cases 
1. imagine we have a large network (or large images) that already by itself needs 11 GB of memory, this can simply crash in the search phase of the nvcaffe, since nvcaffe tries to load multiple algorithms in one workspace and then select the fastest which will need more than the available memory >12GB (= early crash and no training). 

2. Therefore in order to allow the required 'larger combined space for algorithm selection', we end up using a smaller network (or lower resolution images) which can provide the larger  workspace (eg. 7GB)  for nvcaffe, but after the selection phase and cleaning, it will shrink down to the real usage (eg. 4GB) which is completely under-using the available memory for hours of training!

Therefore, there are a couple of strategies than can help:

- A. do the algorithm selection in the sequential fashion, one after the other, which will take couple of more minutes, but will use the adapted amount of ram and not the combined one
- B. let the user choose which algorithm to be used in the proto definition (can we already do this?). in this case we can run a smaller version of a network and see what nvcaffe decides as fastest, and use that flag for all the experiments to come and push the gpu ram to its max.
- C. also as you suggested we can provide a user setting to limit the usage

**Comment by drnikolaev** [comment]
Well, it's a little bit more sophisticated :)
It runs iteration 0 to let all layers set up.
It runs Iteration 1 to let them all allocate their blobs
It runs Iteration 2 to choose winning algos
And only after that it _balances memory between I/O threads and Conv workspaces_.
And now it also releases unused memory. We tested a lot. If model fits in theory, then it fits in practice. If after Iteration 1 you have few megabytes left, it will not even try to go there.
And finally, you have this: `default_conv_algos_override: "1,1,1"`

Also, worth reading: https://docs.nvidia.com/deeplearning/dgx/caffe-user-guide/index.html
Those are NVCaffe-specific settings.

**Comment by RSly** [comment]
Thanks for lights on the details! :)
so to summarize all this, you confirm that from my post above
issue (1) should be already fixed and no crash. nice!
in issue (2), if the network already uses 11GB, we should already see no crash, no algo search, and training starts with one of the algos and 11GB ram usage! 

one last thing: How about one information message from CuDNNConvolutionAlgorithmSeeker that says which algo was decided by nvcaffe? do we have this in the log, so that we can set that specific algo when we want in the future?

**Comment by drnikolaev** [comment]
1. Yes 
2. Yes
3. It's already there:
```
I0521 08:43:56.239918  9311 cudnn_conv_layer.cpp:834] [1] Conv Algos (F,BD,BF): 'conv5' with space 1.79G/2 4 2 5  (limit 11.09G, req 0.34G)
I0521 08:43:56.244078  9313 cudnn_conv_layer.cpp:834] [3] Conv Algos (F,BD,BF): 'conv5' with space 1.79G/2 4 2 5  (limit 11.09G, req 0.34G)
I0521 08:43:56.249747  9316 cudnn_conv_layer.cpp:834] [6] Conv Algos (F,BD,BF): 'conv5' with space 1.79G/2 4 2 5  (limit 11.09G, req 0.34G)
I0521 08:43:56.256482  9310 cudnn_conv_layer.cpp:834] [0] Conv Algos (F,BD,BF): 'conv5' with space 1.79G/2 4 2 5  (limit 11.07G, req 0.34G)
I0521 08:43:56.299268  9310 solver.cpp:327] Iteration 2 (1.136 iter/s, 0.880278s/100 iter), loss = 6.91233
I0521 08:43:56.299288  9310 solver.cpp:349]     Train net output #0: accuracy/top-1 = 0
I0521 08:43:56.299293  9310 solver.cpp:349]     Train net output #1: accuracy/top-5 = 0
I0521 08:43:56.299322  9310 solver.cpp:349]     Train net output #2: loss = 6.91233 (* 1 = 6.91233 loss)
I0521 08:43:56.299695  9312 cudnn_conv_layer.cpp:282] [2] Layer 'conv1' reallocating workspace: 1.79G -> 0.69G
I0521 08:43:56.299695  9311 cudnn_conv_layer.cpp:282] [1] Layer 'conv1' reallocating workspace: 1.79G -> 0.69G
I0521 08:43:56.299789  9313 cudnn_conv_layer.cpp:282] [3] Layer 'conv1' reallocating workspace: 1.79G -> 0.69G
I0521 08:43:56.299796  9314 cudnn_conv_layer.cpp:282] [4] Layer 'conv1' reallocating workspace: 1.79G -> 0.69G
```
This part: `1.79G/2 4 2 5  (limit 11.09G, req 0.34G)` means that current workspace is 1.79G, it serves 2 groups (`/2`), algos `4 2 5` have been chosen, total available device memory is 11.09G and required memory for the algos `4 2 5` is actually 0.34G (per group).
Now, when it's all done it's time to release unused memory:
```
I0521 08:43:56.299695  9312 cudnn_conv_layer.cpp:282] [2] Layer 'conv1' reallocating workspace: 1.79G -> 0.69G
```
You may wonder why 0.69G? This is 0.34G mentioned above multiplied by 2 groups.
Simple and effective.

**Comment by RSly** [comment]
thanks for the confirmation and the explanation,
I was exactly trying to understand those lines! :)
